spark/sql-20/src/itest/scala/org/elasticsearch/spark/integration/AbstractScalaEsSparkSQL.scala

```diff
@@ -2368,6 +2368,46 @@ class AbstractScalaEsScalaSparkSQL(prefix: String, readMetadata: jl.Boolean, pus
     assertEquals(0, result(0).size)
   }
 
+  @Test
+  def testNestedFieldsUpsert(): Unit = {
+    val update_params = "new_samples: samples"
+    val update_script = "ctx._source.samples = params.new_samples"
+    val es_conf = Map(
+      "es.mapping.id" -> "id",
+      "es.mapping.exclude" -> "id",
+      "es.write.operation" -> "upsert",
+      "es.update.script.params" -> update_params,
+      "es.update.script.inline" -> update_script
+    )
+    val sqlContext = new SQLContext(sc)
+    var data = Seq(Row("1", List(Row("hello"), Row("world"))))
```

Comment:
Can we test this with an empty array in the `samples` field?

Reply:
I just tried this. If I use the code as-is, it doesn't write anything. So something gets tripped up when Elasticsearch tries to apply the script:
```
13:49:26.766 [Executor task launch worker for task 0.0 in stage 0.0 (TID 0)] ERROR org.apache.spark.TaskContextImpl - Error in TaskCompletionListener
org.elasticsearch.hadoop.rest.EsHadoopInvalidRequest: org.elasticsearch.hadoop.rest.EsHadoopRemoteException: x_content_parse_exception: [1:74] [script] failed to parse field [params]
{"update":{"_id":"1"}}
{"script":{"source":"ctx._source.samples = params.new_samples","params":{"new_samples":}},"upsert":{"samples":[]}}

	at org.elasticsearch.hadoop.rest.RestClient.checkResponse(RestClient.java:487) ~[elasticsearch-hadoop-mr-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at org.elasticsearch.hadoop.rest.RestClient.execute(RestClient.java:444) ~[elasticsearch-hadoop-mr-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at org.elasticsearch.hadoop.rest.RestClient.execute(RestClient.java:438) ~[elasticsearch-hadoop-mr-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at org.elasticsearch.hadoop.rest.RestClient.execute(RestClient.java:418) ~[elasticsearch-hadoop-mr-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at org.elasticsearch.hadoop.rest.RestClient.bulk(RestClient.java:236) ~[elasticsearch-hadoop-mr-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at org.elasticsearch.hadoop.rest.bulk.BulkProcessor.tryFlush(BulkProcessor.java:215) ~[elasticsearch-hadoop-mr-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at org.elasticsearch.hadoop.rest.bulk.BulkProcessor.flush(BulkProcessor.java:518) ~[elasticsearch-hadoop-mr-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at org.elasticsearch.hadoop.rest.bulk.BulkProcessor.close(BulkProcessor.java:560) ~[elasticsearch-hadoop-mr-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at org.elasticsearch.hadoop.rest.RestRepository.close(RestRepository.java:219) ~[elasticsearch-hadoop-mr-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at org.elasticsearch.hadoop.rest.RestService$PartitionWriter.close(RestService.java:122) ~[elasticsearch-hadoop-mr-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at org.elasticsearch.spark.rdd.EsRDDWriter$$anon$1.onTaskCompletion(EsRDDWriter.scala:74) ~[elasticsearch-spark_2.12-8.1.0-SNAPSHOT-spark30scala212.jar:8.1.0-SNAPSHOT]
	at org.apache.spark.TaskContextImpl.$anonfun$markTaskCompleted$1(TaskContextImpl.scala:124) ~[spark-core_2.12-3.2.0.jar:3.2.0]
	at org.apache.spark.TaskContextImpl.$anonfun$markTaskCompleted$1$adapted(TaskContextImpl.scala:124) ~[spark-core_2.12-3.2.0.jar:3.2.0]
	at org.apache.spark.TaskContextImpl.$anonfun$invokeListeners$1(TaskContextImpl.scala:137) ~[spark-core_2.12-3.2.0.jar:3.2.0]
	at org.apache.spark.TaskContextImpl.$anonfun$invokeListeners$1$adapted(TaskContextImpl.scala:135) ~[spark-core_2.12-3.2.0.jar:3.2.0]
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62) ~[scala-library-2.12.15.jar:?]
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55) ~[scala-library-2.12.15.jar:?]
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49) ~[scala-library-2.12.15.jar:?]
	at org.apache.spark.TaskContextImpl.invokeListeners(TaskContextImpl.scala:135) ~[spark-core_2.12-3.2.0.jar:3.2.0]
	at org.apache.spark.TaskContextImpl.markTaskCompleted(TaskContextImpl.scala:124) ~[spark-core_2.12-3.2.0.jar:3.2.0]
	at org.apache.spark.scheduler.Task.run(Task.scala:147) ~[spark-core_2.12-3.2.0.jar:3.2.0]
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:506) ~[spark-core_2.12-3.2.0.jar:3.2.0]
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1462) [spark-core_2.12-3.2.0.jar:3.2.0]
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:509) [spark-core_2.12-3.2.0.jar:3.2.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_292]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_292]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_292]
```
If instead I write an empty array in DataFrameValueWriter, then when I read the data out the returned DataFrame is kind of useless. It has no columns. If you try do do something like resultDf.select("samples") you get:
```cannot resolve 'samples' given input columns: [];
'Project ['samples]
+- Relation [] ElasticsearchRelation(Map(es.read.field.as.array.include -> samples, es.resource -> nested_fields_upsert_test),org.apache.spark.sql.SQLContext@5cfeb239,None)

org.apache.spark.sql.AnalysisException: cannot resolve 'samples' given input columns: [];
'Project ['samples]
+- Relation [] ElasticsearchRelation(Map(es.read.field.as.array.include -> samples, es.resource -> nested_fields_upsert_test),org.apache.spark.sql.SQLContext@5cfeb239,None)
```
Still trying to track down why that is.

Reply:
OK I added support for empty arrays (and added them to the test). My initial problems writing out empty arrays were due to the fact that they were the first and only thing I was writing, so Elasticsearch did not infer the correct mappings.